File: apps/hello_world/matmul.c
/*
 * matmul.c - Verification tool for matrix multiplication results
 * 
 * This program parses stdin for lines containing matrix multiplication results
 * from threads.c and verifies them by performing the same computation.
 * Much faster than the Python version!
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <signal.h>
#include <stdbool.h>

#define MAX_LINE_LENGTH 1024

// Function to generate a deterministic matrix A based on i,j indices
void generate_matrix_A(int *matrix, int size) {
    for (int i = 0; i < size; i++) {
        for (int j = 0; j < size; j++) {
            matrix[i * size + j] = (i + j) % 100; // A[i,j] = (i + j) % 100
        }
    }
}

// Function to generate a deterministic matrix B based on i,j indices
void generate_matrix_B(int *matrix, int size) {
    for (int i = 0; i < size; i++) {
        for (int j = 0; j < size; j++) {
            matrix[i * size + j] = (i * j + 1) % 100; // B[i,j] = (i * j + 1) % 100
        }
    }
}

// Function to perform matrix multiplication: C = A * B
void matrix_multiply(int *A, int *B, int *C, int size) {
    for (int i = 0; i < size; i++) {
        for (int j = 0; j < size; j++) {
            C[i * size + j] = 0;
            for (int k = 0; k < size; k++) {
                // Use long long to prevent integer overflow
                long long temp = (long long)A[i * size + k] * B[k * size + j];
                C[i * size + j] += (int)(temp % 1000000); // Keep result manageable
            }
        }
    }
}

// Function to calculate matrix multiplication result sum
long long calculate_matrix_sum(int *C, int size) {
    long long sum = 0;
    for (int i = 0; i < size * size; i++) {
        sum += C[i];
    }
    return sum;
}

// Function to verify matrix multiplication and return expected sum
long long verify_matrix_multiplication(int size) {
    int *A = malloc(size * size * sizeof(int));
    int *B = malloc(size * size * sizeof(int));
    int *C = malloc(size * size * sizeof(int));
    
    if (!A || !B || !C) {
        fprintf(stderr, "Failed to allocate memory for %dx%d matrices\n", size, size);
        free(A);
        free(B);
        free(C);
        return -1;
    }
    
    generate_matrix_A(A, size);
    generate_matrix_B(B, size);
    matrix_multiply(A, B, C, size);
    long long sum = calculate_matrix_sum(C, size);
    
    free(A);
    free(B);
    free(C);
    
    return sum;
}

// Function to parse a line for matrix multiplication result pattern
// Returns 1 if pattern found, 0 otherwise
int parse_line(const char *line, int *thread_id, int *size, long long *reported_sum) {
    // Pattern: [thread_id=%d][size=%d][sum=%lld]
    if (sscanf(line, "%*[^[][thread_id=%d][size=%d][sum=%lld]", 
               thread_id, size, reported_sum) == 3) {
        return 1;
    }
    return 0;
}

static volatile sig_atomic_t g_interrupted = 0;

static int g_total_results = 0;
static int g_correct_results = 0;
static int g_incorrect_results = 0;

struct bad_result {
    int thread_id;
    int size;
    long long reported_sum;
    long long expected_sum;
};

static struct bad_result *g_bad = NULL;
static int g_bad_cap = 0;
static int g_bad_len = 0;

static void ensure_bad_capacity(void) {
    if (g_bad_len < g_bad_cap) return;
    int new_cap = g_bad_cap == 0 ? 16 : g_bad_cap * 2;
    struct bad_result *nb = realloc(g_bad, new_cap * sizeof(*nb));
    if (!nb) return; // best-effort; if realloc fails, we just skip recording
    g_bad = nb;
    g_bad_cap = new_cap;
}

static void sigint_handler(int signo) {
    (void)signo;
    g_interrupted = 1;
}

static void print_summary(void) {
    printf("==================================================\n");
    printf("VERIFICATION SUMMARY\n");
    printf("==================================================\n");

    if (g_total_results == 0) {
        printf("No matrix multiplication results found in input.\n");
        printf("Expected format: [thread_id=X][size=Y][sum=Z]\n");
        return;
    }

    printf("Total results found: %d\n", g_total_results);
    printf("Correct results: %d\n", g_correct_results);
    printf("Incorrect results: %d\n", g_incorrect_results);
    printf("Accuracy: %.1f%%\n", g_total_results ? (double)g_correct_results / g_total_results * 100.0 : 0.0);

    if (g_incorrect_results > 0) {
        printf("\nIncorrect results (up to %d shown):\n", g_bad_len);
        for (int i = 0; i < g_bad_len; i++) {
            printf("  Thread %d: Size %dx%d, Reported %lld, Expected %lld, Diff %lld\n",
                   g_bad[i].thread_id, g_bad[i].size, g_bad[i].size,
                   g_bad[i].reported_sum, g_bad[i].expected_sum,
                   g_bad[i].reported_sum - g_bad[i].expected_sum);
        }
    } else {
        printf("\nAll results are correct!\n");
    }
}

int main(int argc, char *argv[])
{
    printf("Matrix Multiplication Verification Tool (C version)\n");
    printf("==================================================\n");

    struct sigaction sa;
    memset(&sa, 0, sizeof(sa));
    sa.sa_handler = sigint_handler;
    sigemptyset(&sa.sa_mask);
    sa.sa_flags = 0;
    sigaction(SIGINT, &sa, NULL);
    
    char line[MAX_LINE_LENGTH];
    
    // Read from stdin line by line
    while (fgets(line, sizeof(line), stdin)) {
        if (g_interrupted) break;

        int thread_id, size;
        long long reported_sum;
        
        // Parse for matrix multiplication results
        if (parse_line(line, &thread_id, &size, &reported_sum)) {
            printf("Found result: Thread %d, Size %dx%d, Reported sum: %lld\n", 
                   thread_id, size, size, reported_sum);
            
            // Verify by performing the same computation
            long long expected_sum = verify_matrix_multiplication(size);
            
            if (expected_sum == -1) {
                printf("  ERROR: Failed to verify computation\n\n");
                g_incorrect_results++;
            } else {
                // Compare results
                int is_correct = (reported_sum == expected_sum);
                const char *status = is_correct ? "CORRECT" : "INCORRECT";
                
                printf("  Expected sum: %lld\n", expected_sum);
                printf("  Status: %s\n", status);
                printf("\n");
                
                if (is_correct) {
                    g_correct_results++;
                } else {
                    g_incorrect_results++;
                    ensure_bad_capacity();
                    if (g_bad_len < g_bad_cap) {
                        g_bad[g_bad_len].thread_id = thread_id;
                        g_bad[g_bad_len].size = size;
                        g_bad[g_bad_len].reported_sum = reported_sum;
                        g_bad[g_bad_len].expected_sum = expected_sum;
                        g_bad_len++;
                    }
                }
            }
            
            g_total_results++;
        }
    }
    
    // Print verification summary
    print_summary();
    
    free(g_bad);
    
    return 0;
}

File: apps/hello_world/matmul.py
#!/usr/bin/env python3
"""
matmul.py - Verification tool for matrix multiplication results

This script parses stdin for lines containing matrix multiplication results
from threads.c and verifies them by performing the same computation.
"""

import sys
import re
import numpy as np

def generate_matrix_A(size):
    """Generate matrix A: A[i,j] = (i + j) % 100"""
    A = np.zeros((size, size), dtype=np.int64)
    for i in range(size):
        for j in range(size):
            A[i, j] = (i + j) % 100
    return A

def generate_matrix_B(size):
    """Generate matrix B: B[i,j] = (i * j + 1) % 100"""
    B = np.zeros((size, size), dtype=np.int64)
    for i in range(size):
        for j in range(size):
            B[i, j] = (i * j + 1) % 100
    return B

def matrix_multiply(A, B, size):
    """Perform matrix multiplication: C = A * B with overflow handling"""
    C = np.zeros((size, size), dtype=np.int64)
    for i in range(size):
        for j in range(size):
            C[i, j] = 0
            for k in range(size):
                # Use int64 to prevent integer overflow
                temp = int(A[i, k]) * int(B[k, j])
                C[i, j] += int(temp % 1000000)  # Keep result manageable
    return C

def calculate_matrix_sum(C):
    """Calculate sum of all elements in matrix C"""
    return np.sum(C)

def verify_matrix_multiplication(size):
    """Perform matrix multiplication and return the expected sum"""
    A = generate_matrix_A(size)
    B = generate_matrix_B(size)
    C = matrix_multiply(A, B, size)
    return calculate_matrix_sum(C)

def parse_line(line):
    """Parse a line for matrix multiplication result pattern"""
    pattern = r'\[thread_id=(\d+)\]\[size=(\d+)\]\[sum=(\d+)\]'
    match = re.search(pattern, line)
    if match:
        thread_id = int(match.group(1))
        size = int(match.group(2))
        reported_sum = int(match.group(3))
        return thread_id, size, reported_sum
    return None

def main():
    """Main verification function"""
    print("Matrix Multiplication Verification Tool")
    print("=" * 50)
    
    results = []
    line_count = 0
    
    # Read from stdin
    for line in sys.stdin:
        line_count += 1
        line = line.strip()
        
        # Parse for matrix multiplication results
        parsed = parse_line(line)
        if parsed:
            thread_id, size, reported_sum = parsed
            print(f"Found result: Thread {thread_id}, Size {size}x{size}, Reported sum: {reported_sum}")
            
            # Verify by performing the same computation
            expected_sum = verify_matrix_multiplication(size)
            
            # Compare results
            is_correct = (reported_sum == expected_sum)
            status = "✓ CORRECT" if is_correct else "✗ INCORRECT"
            
            print(f"  Expected sum: {expected_sum}")
            print(f"  Status: {status}")
            print(f"  Difference: {reported_sum - expected_sum}")
            print()
            
            results.append({
                'thread_id': thread_id,
                'size': size,
                'reported_sum': reported_sum,
                'expected_sum': expected_sum,
                'is_correct': is_correct
            })
    
    # Print verification summary
    print("=" * 50)
    print("VERIFICATION SUMMARY")
    print("=" * 50)
    
    if not results:
        print("No matrix multiplication results found in input.")
        print("Expected format: [thread_id=X][size=Y][sum=Z]")
        return
    
    total_results = len(results)
    correct_results = sum(1 for r in results if r['is_correct'])
    incorrect_results = total_results - correct_results
    
    print(f"Total results found: {total_results}")
    print(f"Correct results: {correct_results}")
    print(f"Incorrect results: {incorrect_results}")
    print(f"Accuracy: {correct_results/total_results*100:.1f}%")
    
    if incorrect_results > 0:
        print("\nIncorrect results:")
        for result in results:
            if not result['is_correct']:
                print(f"  Thread {result['thread_id']}: "
                      f"Size {result['size']}x{result['size']}, "
                      f"Reported {result['reported_sum']}, "
                      f"Expected {result['expected_sum']}, "
                      f"Diff {result['reported_sum'] - result['expected_sum']}")
    
    if correct_results == total_results:
        print("\nAll results are correct!")
    else:
        print(f"\nFound {incorrect_results} incorrect result(s)")

if __name__ == "__main__":
    main()
